src/model_clay.py

```diff
@@ -892,13 +893,25 @@ def predict_step(
         )
         assert not torch.isnan(embeddings_raw).any()  # ensure no NaNs in embedding
 
-        # Take the mean of the embeddings along the sequence_length dimension
-        # excluding the last two latlon_ and time_ embeddings, i.e. compute
-        # mean over patch embeddings only
-        embeddings_mean: torch.Tensor = embeddings_raw[:, :-2, :].mean(dim=1)
-        assert embeddings_mean.shape == torch.Size(
-            [self.model.encoder.B, 768]  # (batch_size, hidden_size)
-        )
+        if self.hparams.output_patch_embeddings:
+            embeddings_raw = rearrange(
+                embeddings_raw[:, :-2, :], "b (w h g) s -> b (w h s) g", w=16, h=16, g=6
```

Comment:
The order of the tensor matters, in this case the ouput from the encoder i.e embeddings_raw is of shape `batch x (group x  num_spatial_patches) x embedding_dims`.

So, the einops operation should be, `b (g l) d -> b g l d` - you can check this notebook for reference: https://github.com/Clay-foundation/model/blob/docs/model/docs/clay-v0-visualization.ipynb

Reply:
Done, but please review again to make sure I got it right this time. Related to this, do you think [this](https://github.com/Clay-foundation/model/pull/133/files#diff-2f39d0012d2540cdab1ac1ff882a3875d829b85f0acee68fe102e6b8a4f2c6c0R171) is a good way to unravel the patch embeddings?

Reply:
Mentioned this below at https://github.com/Clay-foundation/model/pull/133#discussion_r1464236962, but can we just keep the unravelled shape (B, 256, 768), or (B, 1536, 768)? Less work for the downstream user since they won't need to figure out how to unravel the tensor.

Reply:
Yes absolutely. Don't know why I thought it had to be a 1d array...

Reply:
That said, do we want to store a (256, 768) embedding tensor in a single row, or split it into 256 rows of 768-length embeddings? I'm reading the original thread at #127, and it sounds like we need to enable searching on the 32x32 patch level instead of 256x256 chips, which might mean storing a row for each patch?

We can store 2D arrays in GeoParquet, but I'm not sure if vector databases allow indexing 2D arrays, or if we need to make it 1D. Note that this would increase the embedding file size significantly (x256), and the vector database indexing will be much slower. But if these raw embeddings are only meant to be generated on an ad-hoc basis for small locations, and not for similarity search applications, it should be fine.

Reply:
I thought about that too, but that would also imply that we have to track the MGRS tile and the source path 256x more. That seems ineficient. My thinking was that the splitting into rows can happen at the moment on ingestion to a vector search DB or when analysing if required. For "local manual inspection" situations, having the patches in a single array is useful too. So I vote to keep it at one row in the gpd and let the separation be a downstream issue.